File: 560-subarray-sum-equals-k/main.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        
        ans, prefix_sum = 0, 0
        count = {0: 1}

        for num in nums:
            prefix_sum += num
            ans += count.get(prefix_sum - k, 0)
            count[prefix_sum] = count.get(prefix_sum, 0) + 1 

        return ans

File: 560-subarray-sum-equals-k/main.rs
use std::collections::HashMap;


impl Solution {
    pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
        
        let mut ans = 0;
        let mut prefix_sum = 0;

        let mut count: HashMap<i32, i32> = HashMap::new();  
        count.insert(0, 1);

        for num in nums.iter() {
            prefix_sum += num;
            if let Some(&c) = count.get(&(prefix_sum - k)) {
                ans += c;
            }

            *count.entry(prefix_sum).or_insert(0) += 1;
        }


        ans
        
    }
}
